File: src/stockpilot/analysis/charts.py
"""Interactive Plotly-based chart generation for stock analysis.

Generates K-line (candlestick) charts with technical indicator overlays
and backtest equity curve charts. Outputs standalone HTML files.
"""

from __future__ import annotations

import logging
from pathlib import Path
from typing import Any

import pandas as pd

logger = logging.getLogger(__name__)


def create_kline_chart(
    df: pd.DataFrame,
    symbol: str = "",
    title: str | None = None,
    indicators: list[str] | None = None,
    output_path: str | None = None,
    show: bool = False,
) -> str | None:
    """Create an interactive candlestick chart with indicator overlays.

    Args:
        df: DataFrame with columns: date/open/high/low/close/volume + indicator columns
        symbol: Stock symbol for the title
        title: Chart title (defaults to "K-Line: {symbol}")
        indicators: List of indicator columns to overlay (e.g., ["ma_5", "ma_20", "boll_upper"])
        output_path: Path to save HTML file. If None, returns HTML string.
        show: Whether to open in browser (requires display)

    Returns:
        HTML string if output_path is None, else the output file path.
    """
    try:
        import plotly.graph_objects as go
        from plotly.subplots import make_subplots
    except ImportError:
        logger.error("plotly not installed. Run: pip install plotly")
        return None

    if indicators is None:
        indicators = ["ma_5", "ma_20", "ma_60"]

    title = title or f"K-Line: {symbol}"
    dates = df["date"] if "date" in df.columns else df.index

    fig = make_subplots(
        rows=3, cols=1,
        shared_xaxes=True,
        vertical_spacing=0.03,
        row_heights=[0.6, 0.2, 0.2],
        subplot_titles=[title, "Volume", "MACD"],
    )

    # Candlestick
    fig.add_trace(
        go.Candlestick(
            x=dates,
            open=df["open"],
            high=df["high"],
            low=df["low"],
            close=df["close"],
            name="K-Line",
            increasing_line_color="#ef5350",  # Red up (Chinese market convention)
            decreasing_line_color="#26a69a",  # Green down
        ),
        row=1, col=1,
    )

    # Indicator overlays
    colors = ["#FF9800", "#2196F3", "#9C27B0", "#4CAF50", "#F44336", "#00BCD4"]
    for i, ind in enumerate(indicators):
        if ind in df.columns:
            fig.add_trace(
                go.Scatter(
                    x=dates, y=df[ind],
                    mode="lines",
                    name=ind.upper(),
                    line=dict(width=1, color=colors[i % len(colors)]),
                ),
                row=1, col=1,
            )

    # Bollinger Bands as filled area
    if "boll_upper" in df.columns and "boll_lower" in df.columns:
        if "boll_upper" not in indicators:
            fig.add_trace(
                go.Scatter(
                    x=dates, y=df["boll_upper"],
                    mode="lines", name="BOLL Upper",
                    line=dict(width=0.5, color="rgba(156,39,176,0.3)"),
                ),
                row=1, col=1,
            )
            fig.add_trace(
                go.Scatter(
                    x=dates, y=df["boll_lower"],
                    mode="lines", name="BOLL Lower",
                    fill="tonexty",
                    fillcolor="rgba(156,39,176,0.05)",
                    line=dict(width=0.5, color="rgba(156,39,176,0.3)"),
                ),
                row=1, col=1,
            )

    # Volume
    if "volume" in df.columns:
        colors_vol = [
            "#ef5350" if c >= o else "#26a69a"
            for c, o in zip(df["close"], df["open"])
        ]
        fig.add_trace(
            go.Bar(x=dates, y=df["volume"], name="Volume", marker_color=colors_vol),
            row=2, col=1,
        )

    # MACD
    if "macd" in df.columns:
        fig.add_trace(
            go.Scatter(x=dates, y=df["macd"], mode="lines", name="MACD",
                        line=dict(color="#2196F3", width=1)),
            row=3, col=1,
        )
    if "macd_signal" in df.columns:
        fig.add_trace(
            go.Scatter(x=dates, y=df["macd_signal"], mode="lines", name="Signal",
                        line=dict(color="#FF9800", width=1)),
            row=3, col=1,
        )
    if "macd_hist" in df.columns:
        hist_colors = ["#ef5350" if v >= 0 else "#26a69a" for v in df["macd_hist"]]
        fig.add_trace(
            go.Bar(x=dates, y=df["macd_hist"], name="Histogram", marker_color=hist_colors),
            row=3, col=1,
        )

    fig.update_layout(
        template="plotly_dark",
        height=800,
        xaxis_rangeslider_visible=False,
        showlegend=True,
        legend=dict(orientation="h", yanchor="bottom", y=1.02, xanchor="right", x=1),
    )

    if output_path:
        fig.write_html(output_path)
        logger.info("Chart saved to %s", output_path)
        if show:
            fig.show()
        return output_path
    else:
        return fig.to_html(full_html=True, include_plotlyjs="cdn")


def create_equity_chart(
    dates: list[str],
    equity_curve: list[float],
    initial_capital: float = 1_000_000,
    title: str = "Backtest Equity Curve",
    output_path: str | None = None,
    trades: list[dict] | None = None,
) -> str | None:
    """Create a backtest equity curve chart.

    Args:
        dates: List of date strings
        equity_curve: List of equity values over time
        initial_capital: Starting capital (for reference line)
        title: Chart title
        output_path: Path to save HTML file
        trades: Optional list of trade dicts for annotation

    Returns:
        HTML string or file path.
    """
    try:
        import plotly.graph_objects as go
    except ImportError:
        logger.error("plotly not installed")
        return None

    fig = go.Figure()

    fig.add_trace(go.Scatter(
        x=dates, y=equity_curve,
        mode="lines", name="Portfolio Equity",
        fill="tozeroy",
        fillcolor="rgba(33,150,243,0.1)",
        line=dict(color="#2196F3", width=2),
    ))

    fig.add_hline(
        y=initial_capital,
        line_dash="dash",
        line_color="gray",
        annotation_text=f"Initial: ¥{initial_capital:,.0f}",
    )

    # Annotate buy/sell trades
    if trades:
        buys = [t for t in trades if t.get("action") == "buy"]
        sells = [t for t in trades if t.get("action") == "sell"]

        if buys:
            fig.add_trace(go.Scatter(
                x=[t["date"] for t in buys],
                y=[t.get("equity", initial_capital) for t in buys],
                mode="markers", name="Buy",
                marker=dict(symbol="triangle-up", size=10, color="#ef5350"),
            ))
        if sells:
            fig.add_trace(go.Scatter(
                x=[t["date"] for t in sells],
                y=[t.get("equity", initial_capital) for t in sells],
                mode="markers", name="Sell",
                marker=dict(symbol="triangle-down", size=10, color="#26a69a"),
            ))

    fig.update_layout(
        title=title,
        template="plotly_dark",
        height=400,
        yaxis_title="Equity (¥)",
        xaxis_title="Date",
    )

    if output_path:
        fig.write_html(output_path)
        return output_path
    else:
        return fig.to_html(full_html=True, include_plotlyjs="cdn")

File: src/stockpilot/web/static/js/charts.js
/* StockPilot charts */

const CHART_LAYOUT = {
    paper_bgcolor: 'transparent',
    plot_bgcolor: 'rgba(30,41,59,0.5)',
    font: { color: '#94a3b8', size: 11 },
    margin: { t: 30, r: 30, b: 40, l: 50 },
    xaxis: { gridcolor: 'rgba(71,85,105,0.3)', rangeslider: { visible: false } },
    yaxis: { gridcolor: 'rgba(71,85,105,0.3)' },
    legend: { orientation: 'h', y: 1.08, x: 0.5, xanchor: 'center', font: { size: 10 } },
};

const PLOTLY_CONFIG = { responsive: true, displayModeBar: true, modeBarButtonsToRemove: ['lasso2d', 'select2d'] };
const SERIES_COLORS = ['#3b82f6', '#ef4444', '#10b981', '#f59e0b', '#8b5cf6', '#06b6d4'];

function createKlineChart(container, data, symbol) {
    const dates = data.map(d => d.date);
    const traces = [
        {
            type: 'candlestick',
            x: dates,
            open: data.map(d => d.open),
            high: data.map(d => d.high),
            low: data.map(d => d.low),
            close: data.map(d => d.close),
            name: 'K线',
            increasing: { line: { color: '#ef5350' } },
            decreasing: { line: { color: '#26a69a' } },
            xaxis: 'x',
            yaxis: 'y',
        },
    ];

    const maColors = { ma_5: '#f59e0b', ma_10: '#3b82f6', ma_20: '#8b5cf6', ma_60: '#22c55e' };
    for (const [key, color] of Object.entries(maColors)) {
        const vals = data.map(d => d[key] ?? null);
        if (vals.some(v => v != null)) {
            traces.push({
                type: 'scatter',
                mode: 'lines',
                x: dates,
                y: vals,
                name: key.toUpperCase().replace('_', ''),
                line: { width: 1, color },
                xaxis: 'x',
                yaxis: 'y',
            });
        }
    }

    const bollUpper = data.map(d => d.boll_upper ?? null);
    const bollLower = data.map(d => d.boll_lower ?? null);
    if (bollUpper.some(v => v != null)) {
        traces.push({
            type: 'scatter',
            mode: 'lines',
            x: dates,
            y: bollUpper,
            name: 'BOLL上',
            line: { width: 0.7, color: 'rgba(168,85,247,0.45)', dash: 'dot' },
            xaxis: 'x',
            yaxis: 'y',
        });
        traces.push({
            type: 'scatter',
            mode: 'lines',
            x: dates,
            y: bollLower,
            name: 'BOLL下',
            line: { width: 0.7, color: 'rgba(168,85,247,0.45)', dash: 'dot' },
            fill: 'tonexty',
            fillcolor: 'rgba(168,85,247,0.04)',
            xaxis: 'x',
            yaxis: 'y',
        });
    }

    traces.push({
        type: 'bar',
        x: dates,
        y: data.map(d => d.volume),
        name: '成交量',
        marker: { color: data.map(d => d.close >= d.open ? '#ef5350' : '#26a69a') },
        xaxis: 'x2',
        yaxis: 'y2',
    });

    if (data.some(d => d.macd != null)) {
        traces.push({
            type: 'scatter',
            mode: 'lines',
            x: dates,
            y: data.map(d => d.macd ?? null),
            name: 'MACD',
            line: { width: 1, color: '#3b82f6' },
            xaxis: 'x3',
            yaxis: 'y3',
        });
        traces.push({
            type: 'scatter',
            mode: 'lines',
            x: dates,
            y: data.map(d => d.macd_signal ?? null),
            name: 'Signal',
            line: { width: 1, color: '#f59e0b' },
            xaxis: 'x3',
            yaxis: 'y3',
        });
        traces.push({
            type: 'bar',
            x: dates,
            y: data.map(d => d.macd_hist ?? null),
            name: 'Hist',
            marker: { color: data.map(d => (d.macd_hist ?? 0) >= 0 ? '#ef5350' : '#26a69a') },
            xaxis: 'x3',
            yaxis: 'y3',
        });
    }

    Plotly.newPlot(container, traces, {
        ...CHART_LAYOUT,
        height: 620,
        grid: { rows: 3, columns: 1, subplots: [['xy'], ['x2y2'], ['x3y3']], roworder: 'top to bottom' },
        xaxis: { ...CHART_LAYOUT.xaxis, domain: [0, 1], anchor: 'y' },
        yaxis: { ...CHART_LAYOUT.yaxis, domain: [0.42, 1], title: '价格' },
        xaxis2: { ...CHART_LAYOUT.xaxis, domain: [0, 1], anchor: 'y2', showticklabels: false },
        yaxis2: { ...CHART_LAYOUT.yaxis, domain: [0.22, 0.38], title: '成交量' },
        xaxis3: { ...CHART_LAYOUT.xaxis, domain: [0, 1], anchor: 'y3' },
        yaxis3: { ...CHART_LAYOUT.yaxis, domain: [0, 0.18], title: 'MACD' },
        title: { text: `${symbol} K线工作台`, font: { size: 14 } },
    }, PLOTLY_CONFIG);
}

function createLineComparisonChart(container, series, title, yTitle = '数值') {
    const traces = series.map((item, idx) => ({
        type: 'scatter',
        mode: 'lines',
        x: item.x,
        y: item.y,
        name: item.name,
        line: { width: 2, color: item.color || SERIES_COLORS[idx % SERIES_COLORS.length] },
    }));

    Plotly.newPlot(container, traces, {
        ...CHART_LAYOUT,
        height: 360,
        title: { text: title, font: { size: 14 } },
        yaxis: { ...CHART_LAYOUT.yaxis, title: yTitle },
    }, PLOTLY_CONFIG);
}

function createEquityCurve(container, dates, equity, initialCapital) {
    return createLineComparisonChart(
        container,
        [{ name: '组合净值', x: dates, y: equity, color: '#3b82f6' }],
        '资金曲线',
        '净值 (¥)',
    );
}

function createEquityComparison(container, runs, initialCapital) {
    const series = runs.map((run, idx) => ({
        name: run.label || `${run.symbol} · ${run.strategy}`,
        x: run.dates,
        y: run.equity_curve,
        color: SERIES_COLORS[idx % SERIES_COLORS.length],
    }));
    return createLineComparisonChart(container, series, '多组回测对比', '净值 (¥)');
}

function createPieChart(container, labels, values, title) {
    Plotly.newPlot(container, [{
        type: 'pie',
        labels,
        values,
        hole: 0.45,
        textinfo: 'label+percent',
        marker: { colors: SERIES_COLORS },
        textfont: { color: '#e2e8f0', size: 11 },
    }], {
        ...CHART_LAYOUT,
        height: 320,
        title: { text: title || '', font: { size: 13 } },
    }, PLOTLY_CONFIG);
}
